rpc/src/endpoint/consensus_state.rs

```diff
@@ -0,0 +1,387 @@
+//! `/consensus_state` endpoint JSON-RPC wrapper
+
+use crate::{Error, Method};
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::convert::TryFrom;
+use std::fmt;
+use std::str::FromStr;
+use subtle_encoding::hex;
+use tendermint::block::{Height, Round};
+use tendermint::{account, hash, vote, Hash, Time};
+
+// From https://github.com/tendermint/tendermint/blob/e820e68acd69737cfb63bc9ccca5f5450a42b5cf/types/vote.go#L16
+const NIL_VOTE_STR: &str = "nil-Vote";
+
+/// Get the current consensus state.
+#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
+#[non_exhaustive]
+pub struct Request;
+
+impl Request {
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl crate::Request for Request {
+    type Response = Response;
+
+    fn method(&self) -> Method {
+        Method::ConsensusState
+    }
+}
+
+impl crate::SimpleRequest for Request {}
+
+/// The current consensus state (UNSTABLE).
+///
+/// Currently based on https://github.com/tendermint/tendermint/blob/e820e68acd69737cfb63bc9ccca5f5450a42b5cf/consensus/types/round_state.go#L97
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct Response {
+    pub round_state: RoundState,
+}
+
+impl crate::Response for Response {}
+
+/// The state of a particular consensus round.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct RoundState {
+    #[serde(alias = "height/round/step")]
+    pub height_round_step: HeightRoundStep,
+
+    #[serde(with = "tendermint::serializers::time")]
+    pub start_time: Time,
+
+    #[serde(with = "hash::allow_empty")]
+    pub proposal_block_hash: Hash,
+
+    #[serde(with = "hash::allow_empty")]
+    pub locked_block_hash: Hash,
+
+    #[serde(with = "hash::allow_empty")]
+    pub valid_block_hash: Hash,
+
+    pub height_vote_set: Vec<RoundVotes>,
+
+    pub proposer: ValidatorInfo,
+}
+
+/// A compound object indicating a height, round and step for consensus state.
+#[derive(Clone, Debug)]
+pub struct HeightRoundStep {
+    /// Current block height
+    pub height: Height,
+    /// Current consensus round
+    pub round: Round,
+    /// Current consensus step
+    pub step: i8,
+}
+
+impl Serialize for HeightRoundStep {
```

Comment:
I know we deprioritized separating JSON serialization in the code, but would it make sense to add one-off JSON structs when we implement new domain types?

This seems like a prime example where we can create a HeightRoundStep JSON type that has i64/i64/i8 fields and then add a `#[serde(try_from="json-type", into="json-type")]` to the top of this struct.

I'm asking because tracking custom serialization implementations can become tricky and this implementation has public fields (described as a Type 0 struct in issue #661 ) but domain knowledge restrictions in the serialization (which I consider no-no and should be either in a TryFrom trait or a constructor as described by Type 1 and Type 2 structs in the same issue). I would guess that this doesn't cause issues in regular cases, because most of the fields are well guarded (Height and Round are domain types), but I can for example create a HeightRoundStep with step == -2 which I think is invalid.

So, even if we keep the custom serialization, it seems that a constructor might be warranted.

Reply:
We could separate the JSON struct from the domain type for `HeightRoundStep`, but where would we put it?

Reply:
Chatted to @greg-szabo about this, and we'll follow up here in a separate issue (#749).